### ISSUE TITLE
Mainly Saving aspects from the new Log/Script window, also backing up the data

### ISSUE THREAD

**Issue by rdstern** [africanmathsinitiative/R-Instat#8357]
@lloyddewit I wondered what happens elsewhere and here is information [about RStudio](http://www2.hawaii.edu/~georgeha/Handouts/meas/_book/saving.html)

It says:

![image](https://github.com/africanmathsinitiative/R-Instat/assets/11226469/888ede11-7e69-4e9d-bf0d-1eaced04d3cf)

Now we currently don't seem to indicate whether a script has been saved or not.  Here is an example:

![image](https://github.com/africanmathsinitiative/R-Instat/assets/11226469/4716f14a-90a6-407c-b890-daebeda43b66)

Here it is obvious, from the name, that Untitled3 has not been saved.  But I loaded test2 and then changed it.  There was no indication.

a) I suggest at least an asterisk after scripts that have not been saved.  That's neat and consistent with RStudio.  Optionally the name could also be coloured?  We already use red for filters and selects, which might mean that red is good - and it is consistent with RStudio.  Or perhaps green could be used for scripts that have been saved? Or just let the asterisk be the indication?

b) The Save button is currently a Save As button.  It asks every time.  I suggest there should be a Save Button.  So, it asks if there is no name, but if a file was loaded, or a script was saved previously, then it doesn't ask.  The asterisk would just disappear.  I would be happy for this button to become the Save Button.  If you want the Save As, then it is on the File > Save As > Save Script As.  

 That leaves us with the Log.  It used to be a special log window and not it is (just?) the first tab.  @Patowhiz I have a more major suggestion for the log tab.

c) Our default is to backup avery 10 minutes.  I think this is a backup of both the workbook and the log.  The log is quite small.  Could we update the log after every command?  So it is always up-to-date?  (Of course it is in an R-Instat place, not a place defined by the user.)

d) So I suggest two possibilities here.  First is that, at any time, in a session, you may opt to Save the log file.  That's just like saving any other log/script tab.  If you do that, then that named file is then updated, rather than the secret R-Instat location file.

e) On startup next time, and presumably also somewhere in the file menu you may choose to re-open the previous log file.  Maybe that is an entry in brackets.  Maybe it comes under the help, and maybe it could be in the previous files.  Maybe there could be a section for Recent Log/Script files.  Often, for beginners, this might be just `Last log file`?

![image](https://github.com/africanmathsinitiative/R-Instat/assets/11226469/946bf607-317f-4d58-a2c6-9210398c6832)

f) I note, as I was playing with the log-script window, that when I run a command from the log window it (of course) adds again to the log window.  It also adds line breaks, between each line, which seem unnecessary.

g) One reason for emphsising the log window is that it may take the pressure of the saving so much (by default) from the output tabs.  They could become quite big files and usually not always with everything useful.  And we can always regenerate the output from the log file.  And that is a pretty professional way to work.

h) @Patowhiz the suggestion above will also affect the (default 10 minute) backing up procedure.  I suggest this may be another area where we could substantially improve.  With a large data file the backing up can take a considerable time.  I suggest a feature of large data files is that often the data does not change.  It is often the meta-data (adding filters), etc that changes?  When the data has not changed, perhaps the backup can become quicker?

**Comment by rdstern** [comment]
@lloyddewit an additional, very minor change is to change the `Add` label which gives a New tab, into `New`? In the Output Tabs there is both `Add to Existing`, and `Add to New`.   I suggest the New tab in the Scripts is the equivalent of the Add to New Tab in the Output Window.

**Comment by rdstern** [comment]
@Patowhiz and @lloyddewit After discussion with various people, I hope we can progress quickly to a few quick and hopefully uncontroversial wins here?

This has been helped by an improvement in the toolbar.

a) Add the asterisk to all tabs that could be saved in the Log/Script windows and in the Output window.
b) I hope this will make it easy, or automatically add an asterisk to the corresponding entry in Save icon on the toolbar.
c) Change the Save button in the Log/Script and the Output windows to Save, rather than Save As.  I assume this will change the toolbar save options to follow-suit.  (If a Save As is still needed, then it is on the File menu.)
d) Slightly different, but change the `Add` button on the Script window to `New`.
e) Check the feature in the log window that currently leaves a blank line when part of the log window is copied and then run.

I am leaving the discussion of the backing up and the possible use of the log file for later.  My reasoning is that the priority is a simple system for beginners.  And I think the above is sufficient for now.  Most will need to save the data and may also want to save and copy results from the Output Window.

Saving the log file is a luxury, but will now be easilt available from the toolbar as well as the File menu.

**Comment by lloyddewit** [comment]
On 15/06/23, we (@volloholic @ChrisMarsh82 @Patowhiz @lilyclements and I) discussed consistency between the script and output windows. This is related to this issue. Below are some rough personal notes. If I missed/misunderstood anything, then please let me know. Thanks.

----

- There are some visual inconsistencies between the two windows. As an early task, we should make the two windows use the same font and syntax highlighting.
- The output window is divided up into components which typically hold a snippet of code or an output (e.g. a graph). In WinForms, if we have too many components open, then we get performance issues.
- Each component stores metadata. The metadata includes the code snippet that generated the output.
- Currently, there is no link between the dialog and the output/script windows. We should associate the metadata with the output/script windows.- 
- We don't have a direct equivalent of the RStudio console window (console combines a log of executed R statements, with display of warnings and output).
- We could potentially store the metadata in the long window. But if we did this, then how would we store metadata about scripts in the script window that are not yet executed?

**Comment by rdstern** [comment]
@Patowhiz and @ChrisMarsh82  - can we still consider whether a "save project" can be implemented in the near future.  Patrick, you wrote above:

"I suggest exploring having a save option that saves all the data, log and outputs in a "workspace" folder. Then a user can just load back the "workspace". This is now very possible considering the changes we have been making and we can optimize this process. The workspace can include saved and unsaved items."

David was saying that it may be relatively easy to be able to backup the databook and add the output windows, (including graphs) and log and script windows?  Only for restoring back into R-Instat on a later occasion - and into the same version of R-Instat.  This is not for long-term storage.  It would lose the settings in dialogs and the last commands, etc.

I am trying to consider an update, not too late this autumn, which has key additions.  (And might also be the last version where we cater for 32 bit machines.)  I would really like this facility to be implemented.

**Comment by Patowhiz** [comment]
Yes, with the additions made. That's all possible as a workspace. We can literally restore all the contents of the output and data window if we wanted. Not much design thinking required maybe implementation. 

@rdstern you are now saying this is one of your 32 bits final requirements.
